### ISSUE TITLE
Find all faces for a planar graph

### ISSUE THREAD

**Issue by Ganesh-Nunnagoppula** [networkx/networkx#2807]
I thought it would be good to have a functionality to find all the faces in a planar graph. The current `networkx.algortihms.cycles` doesn't support planar graphs with node positions known. I want to write an additional function like `networkx.algorithms.cycles.planar_faces`. 

I can work on this if the above fuctionality seems good.

**Comment by dschult** [comment]
What doesn't the current cycles routine support? I think it works fine with planar graphs with or without node positions. It just ignores that information. What is it that you want/need that the current code doesn't provide?

**Comment by Ganesh-Nunnagoppula** [comment]
For finding the faces of a planar graph, the `cycle_basis` fails in the following cases

1. If faces are shortest cycles. It fails if some path has lot of nodes.
<img width="267" alt="screen shot 2018-01-16 at 10 31 44 am" src="https://user-images.githubusercontent.com/10352774/35005562-a576d748-faa8-11e7-8a60-a92a40dfd1fb.png">
source: produced by @mjstevens777

2. Faces depend on the postion of the nodes. The following two graphs have the same node connectivity with different node positions resulting in different faces.
<img width="391" alt="screen shot 2018-01-16 at 10 52 41 am" src="https://user-images.githubusercontent.com/10352774/35006556-bba91d7a-faab-11e7-9bbf-78da8acd2c17.png">
<img width="412" alt="screen shot 2018-01-16 at 10 53 00 am" src="https://user-images.githubusercontent.com/10352774/35006563-be0306da-faab-11e7-9fae-490d3b76da6a.png">
source: reporduced in networkx from https://math.stackexchange.com/questions/8140/find-all-cycles-faces-in-a-graph

**Comment by Ganesh-Nunnagoppula** [comment]
@dschult I wanted to check with you regarding the above issue. Do you think the above is a reasonable functionality to have? I wrote the above functionality and tested it locally. I can raise a pull request integrating with NetworkX if you approve of the issue.

**Comment by dschult** [comment]
I don't understand the definitions being used in the examples provided. You say we need to know the positions of the nodes to find the faces. So how would this functionality work then? Do you provide a position dict as input?  How is a face even defined if it requires positions?

**Comment by Ganesh-Nunnagoppula** [comment]
Thanks for the prompt reply. I am sorry my comment above wasn't very informative. 

- The Graph is assumed to be planar meaning no two edges cross each other. To test whether a graph is planar or not we need the postion of nodes 
- The `face` of a planar graph is defined as the regions bounded by a set edges which contain no other vertex or edge.
- A planar graph will satisfy the Euler formula `v - e + f = 2` `v`, `e`, `f` are the number of vertices(nodes), edges and faces. 
- I found a similar issue https://github.com/networkx/networkx/pull/1701
- The functionality I am thinking is `networkx.algorithms.cycles.planar_faces(graph, pos)`
```python
def planar_faces(G, pos):
    """Find all faces of a planar graph.

    Parameters
    ----------
    G : a networkX Graph
        graph is supposed to be planar
    pos : dictionary
        A dictionary with nodes as keys and positions as values.
        example : {'1': array([x, y]),.....} 

    Returns
    -------
    faces : list of lists
        each list consists of nodes of a face in cyclic order.

    Raises
    ------
    NetworkX Error if G is not a planar graph.
        
    Notes
    -----
    Graph:

    1------2------3
    |      |      |
    |      |      |
    4------5------6
    |      |      |
    |      |      |
    7------8------9

    faces : [
        [4, 5, 2, 1],
        [5, 6, 3, 2],
        [7, 8, 5, 4],
        [8, 9, 6, 5],
        [7, 4, 1, 2, 3, 6, 9, 8],  # perimeter 
    ]

    observe that v - e + f = 2, 9(nodes) - 12(edges) + 5(faces) = 2
    """
    pass
```

**Comment by dschult** [comment]
But a planar graph doesn't have coordinates associated with the nodes. The definition is only that there exists such an embedding into the plane. Indeed there are infinitely many such embeddings. All of them satisfy the v-e+f=2. But what does it mean to return the faces? Its seems like cycle_basis does a pretty good job. 

Do you know of applications/extensions that would use the faces?  Would those be able to use the cycle_basis?

**Comment by Ganesh-Nunnagoppula** [comment]
- We use networkX for roof processing. One of the example I provided is a roof, wrongly annotated by a user as a result one of the edge too many nodes along it. This results in giving wrong faces when we use `cycle_basis` see `basis 2` in the 1st example. So we wrote the functionality to find all faces of a given a planar embedding(roof). 
- Another application we came across is while implementing papers like [Polyfit](http://openaccess.thecvf.com/content_iccv_2017/html/Nan_PolyFit_Polygonal_Surface_ICCV_2017_paper.html) we need to identify all faces in a particular plane.
- The algorithm I implemented is from https://mathoverflow.net/questions/23811/reporting-all-faces-in-a-planar-graph. This is similar to representing the planar embedding by a half-edge data structure I think.
